### ISSUE TITLE
ForbiddenMethodCall - IllegalArgumentException exception when running detekt gradle task

### ISSUE THREAD

**Issue by DcortezMeleth** [detekt/detekt#3546]
## Expected Behavior
I would expect no exceptions thrown when detekt runs.

## Observed Behavior
After upgrading detect (gradle plugin) from 1.15.0 to 1.16.0 Exception as below is thrown when analysing one of the classes.
```
The original exception message was: declarationDescriptor is null for constructor = {Comparable<{TwoValueEnum & String}> & java.io.Serializable} with class org.jetbrains.kotlin.types.IntersectionTypeConstructor
Running detekt '1.16.0' on Java '11.0.2+9' on OS 'Mac OS X'
If the exception message does not help, please feel free to create an issue on our GitHub page.
...
Caused by: java.lang.IllegalArgumentException: declarationDescriptor is null for constructor = {Comparable<{HangingWriteStatus & String}> & java.io.Serializable} with class org.jetbrains.kotlin.types.IntersectionTypeConstructor
	at org.jetbrains.kotlin.js.descriptorUtils.DescriptorUtilsKt.getJetTypeFqName(descriptorUtils.kt:23)
	at io.gitlab.arturbosch.detekt.rules.style.ForbiddenMethodCall.check(ForbiddenMethodCall.kt:81)
```

## Steps to Reproduce
I tried to add test case with code that throws exception (code below), only mocking some and simplifying some stuff (not including method params) but in tests everything works fine. 

I managed to discover that issue lies in the `tracer.trace` call second parameter, which is `Map<String,Any>`. It is enough to comment out any entry from this map and exception is not present then. Trick like specifying map verbosely in a manner of `mapOf<String,Void>(...)` or `mapOf(...) as Map<String,Void>` don't help. The same as moving map creation to separate method with return type of `Map<String,Void>`. 
Test case always deduce that type of second parameter is `kotlin.collections.Map`.

Is there something that may differ between running this in real project and test case? I use some machine, IDE (cli gives same error), kotlin and jdk versions.
```
enum class TwoValueEnum(val first: Boolean, val second: Boolean) {
        ONE(false, false),
        TWO(true, true);
    } 
    
    data class TransactionId(private val id: String) {
        fun asString() = id
    }

    class Tracer {
        fun doSth(s: String) {}
    }

    fun <T> Tracer.trace(operationName: String, params: Map<String, Any> = emptyMap(), operation: () -> T): T {
        doSth(operationName)
        print(params)
        return operation()
    }

    class CC(
        private val tracer: Tracer
    ) {
        fun trace(s: String, t: TransactionId, status: TwoValueEnum) =
            tracer.trace(
                "op", 
                mapOf(
                    "a" to t,
                    "x" to status,
                    "y" to s
                )
            ) {
                print("it")
            }
    }
```

## Context
Now funny part. This exception is thrown by ForbiddenMethodCall extension I did lately. After upgrade I wanted to use it in our project to I could remove few Suppresses. In most of the modules it worked as expected but in one of them detekt threw exception instead of generating report. 

As this error was introduce by me I would be happy to try to solve this issue.
Unfortunately for now as I'm not able to reproduce it in test cases and don't know how to debug detekt run from gradle plugin my only idea is to blind fix it, build it locally and test in our project. If that is possible I would prefer to be able to add test cases which will be failing before the fix and pass after so everyone can see it too.

So can you please guide me a bit here? I guess what I would need to know now is if I can debug detekt gradle plugin in other way than logging and also what can the differences be that almost identical code throws exception is project but works fine in a test case. 

## Your Environment

* Version of detekt used: 1.16.0
* Version of Gradle used (if applicable): 6.4
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: Mac OS X 11.2.3
* Link to your project (if it's a public repository):

**Comment by cortinico** [comment]
> So can you please guide me a bit here? I guess what I would need to know now is if I can debug detekt gradle plugin in other way than logging and also what can the differences be that almost identical code throws exception is project but works fine in a test case.

I'm unsure what exactly is happening here. That's the function you're hitting:

https://github.com/JetBrains/kotlin/blob/29b23e79f32791e456a5b4a453277f0f0b3e984d/js/js.frontend/src/org/jetbrains/kotlin/js/descriptorUtils/descriptorUtils.kt#L22-L41

Could we create a reproducer project on GH with your `TwoValueEnum` class that we can investigate?

> Is there something that may differ between running this in real project and test case? I use some machine, IDE (cli gives same error), kotlin and jdk versions.

I believe this could be related to the Kotlin version.
Out of curiosity, could you try with `1.16.0-RC3` as well and let us know if there is an issue?

**Comment by DcortezMeleth** [comment]
Sorry it took such a long time.
I have found and cause of the issue and managed to produce fix.
Problem here is that in some rare cases kotlin complier could infer argument type to be Intersection type. Unfortunately as those types do not have a constructor, `DescriptorUtilsKt.getJetTypeFqName` thrown exception for them.

Problem here is that behaviour seems to depend on something more than plain code that trows it, as copying code fragment responsible for exception didn't cause the same issue. What I mean is that in the original project inferred param type was Intersection but when copy-pasted to other places it returned Any. 
I have also tried other approaches to enforce Intersection types as a param in a test case but was unable to do so.

Luckily while debugging plugin on local project I have managed to come up with two possible solutions.

First one changes way in which we recover parameter types. So instead of calling `resolvedCall.resultingDescriptor.valueParameters` we call `resolvedCall.candidateDescriptor.valueParameters`.
This way we recover parameter types from method declarations instead of runtime.
This has its cons and pros then. The main advantage is that we will never encounter Intersection type here as you cannot instantiate it on your own, so we are save from exception I was getting. The con is that for generic methods like `public infix fun <A, B> A.to(that: B): Pair<A, B> = Pair(this, that)` called like `"string" to "string"` we will be getting argument type of `B` instead of `String`. However I don't think this is a big issue as I don't think scenario where you would like to forbid generic method call with one param type is a relevant one.

Other approach would be to simply catch IllegalArgumentException while getting parameter types and in such case simply skip checked method. It's rather rare scenario so this should be fine too, although it doesn't look as clean as first solution.

I have already open PR with first solution (https://github.com/detekt/detekt/pull/3626) as I will it is cleaner.

I know it does not have a test case to prove it helps, but as much as i wanted to provide ones I was unable to do so. If by a chance you know how to reliably force Intersection type as a param in test case let me know. I will be happy to prove my fix working.
